File: 프로그래머스/lv3/43164. 여행경로/여행경로.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Solution {
    
    static List<String> list = new ArrayList<>();
	static boolean[] visited;
    
    public String[] solution(String[][] tickets) {
        String[] answer = {};
        
        visited = new boolean[tickets.length];
		
		dfs(tickets, 0, "ICN", "ICN");
		
		Collections.sort(list);
		
		answer = list.get(0).split(" ");
        
        return answer;
    }
    
    public static void dfs(String[][] tickets, int depth, String node, String path) {
		if(depth == tickets.length) {
			list.add(path);
		}
		
		for(int i=0; i<tickets.length; i++) {
			
			if(tickets[i][0].equals(node) && visited[i] == false) {
				visited[i] = true;
				dfs(tickets, depth+1, tickets[i][1], path + " " + tickets[i][1]);
				visited[i] = false;
			}
			
		}
		
	}
    
}

File: 프로그래머스/3/43164. 여행경로/여행경로.py
answer = []
def solution(tickets):

    visited = [False] * len(tickets)

    DFS(tickets, ["ICN"],  0, visited)

    return answer

def DFS(tickets, arr, depth, visited):
    global answer
    if depth == len(tickets):
        if not answer or arr < answer:
            answer = arr[:]
        return

    start = arr[depth]

    for j in range(len(tickets)):
        temp = tickets[j]
        if start == temp[0] and not visited[j]:
            visited[j] = True
            DFS(tickets, arr + [temp[1]], depth + 1, visited)
            visited[j] = False